org.eclipse.winery.topologymodeler.ui/src/app/canvas/canvas.component.ts

```diff
@@ -0,0 +1,1403 @@
+                this.artifactName = null;
+                this.artifactType = null;
+                this.deploymentArtifactModal.hide();
+                break;
+        }
+    }
+
+    /**
+     * New nodes can be dragged directly from the palette,
+     * adds the node to the internal representation
+     * @param event  The html event.
+     */
+    moveNewNode(event): void {
+        this.newNodeData = {
+            id: this.newNode.id,
+            x: event.clientX - 108,
+            y: event.clientY - 30
+        };
+        this.allNodeTemplates[this.indexOfNewNode].x = this.newNodeData.x;
+        this.allNodeTemplates[this.indexOfNewNode].otherAttributes.x = this.newNodeData.x;
+        this.allNodeTemplates[this.indexOfNewNode].y = this.newNodeData.y;
+        this.allNodeTemplates[this.indexOfNewNode].otherAttributes.y = this.newNodeData.y;
+    }
+
+    /**
+     * Repositions the new node and repaints the screen
+     * @param $event  The html event.
+     */
+    positionNewNode(event): void {
+        this.updateSelectedNodes('Position new Node');
+        this.unbindNewNodeMouseMove();
+        this.unbindNewNodeMouseUp();
+        setTimeout(() => this.newJsPlumbInstance.revalidate(this.newNode.id), 1);
+        this.repaintJsPlumb();
+    }
+
+    /**
+     * Setter for PaletteState
+     * @param currentPaletteOpened
+     */
+    setPaletteState(currentPaletteOpened: boolean): void {
+        if (currentPaletteOpened) {
+            this.currentPaletteOpenedState = currentPaletteOpened;
+        }
+    }
+
+    /**
+     * Gets called if relationships get deleted or created and calls the
+     * correct handler.
+     * @param currentRelationships  List of all displayed relationships.
+     */
+    updateRelationships(currentRelationships: Array<TRelationshipTemplate>): void {
+        if (currentRelationships.length > 0) {
+            if (currentRelationships.length !== this.allRelationshipTemplates.length) {
+                const difference = currentRelationships.length - this.allRelationshipTemplates.length;
+                if (difference === 1) {
+                    this.handleNewRelationship(currentRelationships);
+                } else if (difference < 1) {
+                    this.handleDeletedRelationships(currentRelationships);
+                } else {
+                    try {
+                        this.handleLoadedRelationships(currentRelationships);
+                    } catch (e) {
+                    }
+                }
+            } else {
+                this.updateRelName(currentRelationships);
+            }
+        }
+    }
+
+    /**
+     * Handler for new relations, adds it to the internal representation
+     * @param currentRelationships  List of all displayed relations.
+     */
+    handleNewRelationship(currentRelationships: Array<TRelationshipTemplate>): void {
+        const newRel = currentRelationships[currentRelationships.length - 1];
```

Comment:
otherwise there would be jsplumb problems with the representation in the dom. this is save because  we have a redux reducer action defined which pushes new nodes and relationships at the end of the corresponding array in the store